autopush-common/src/db/bigtable/bigtable_client/merge.rs

```diff
@@ -0,0 +1,486 @@
+use std::cell::RefCell;
+use std::collections::{BTreeMap, HashMap};
+use std::time::{Duration, SystemTime};
+
+use futures::StreamExt;
+use google_cloud_rust_raw::bigtable::v2::bigtable::{ReadRowsResponse, ReadRowsResponse_CellChunk};
+use grpcio::ClientSStreamReceiver;
+
+use super::{cell::Cell, error::BigTableError, row::Row, FamilyId, Qualifier, RowKey};
+
+/// List of the potential states when we are reading each value from the
+/// returned stream and composing a "row"
+#[derive(Debug, Default, Clone, Eq, PartialEq)]
+enum ReadState {
+    #[default]
+    RowStart,
+    CellStart,
+    CellInProgress,
+    CellComplete,
+    RowComplete,
+}
+
+/// An in-progress Cell data struct.
+#[derive(Debug, Clone)]
+pub(crate) struct PartialCell {
+    pub(crate) family: FamilyId,
+    /// A "qualifier" is a column name
+    pub(crate) qualifier: Qualifier,
+    /// Timestamps are returned as microseconds, but need to be
+    /// specified as milliseconds (even though the function asks
+    /// for microseconds, you * 1000 the mls).
+    pub(crate) timestamp: SystemTime,
+    /// Not sure if or how these are used
+    pub(crate) labels: Vec<String>,
+    /// The data buffer.
+    pub(crate) value: Vec<u8>,
+    /// the returned sort order for the cell.
+    pub(crate) value_index: usize,
+}
+
+impl Default for PartialCell {
+    fn default() -> Self {
+        Self {
+            family: String::default(),
+            qualifier: String::default(),
+            timestamp: SystemTime::now(),
+            labels: Vec::new(),
+            value: Vec::new(),
+            value_index: 0,
+        }
+    }
+}
+
+/// An in-progress Row structure
+#[derive(Debug, Default, Clone)]
+pub(crate) struct PartialRow {
+    /// table uniquie key
+    row_key: RowKey,
+    /// map of cells per family identifier.
+    cells: HashMap<FamilyId, Vec<PartialCell>>,
+    /// the last family id string we encountered
+    last_family: FamilyId,
+    /// the working set of family and cells
+    /// we've encountered so far
+    last_family_cells: HashMap<FamilyId, Vec<Cell>>,
+    /// the last column name we've encountered
+    last_qualifier: Qualifier,
+    /// Any cell that may be in progress (chunked
+    /// across multiple portions)
+    cell_in_progress: RefCell<PartialCell>,
+}
+
+/// workhorse struct, this is used to gather item data from the stream and build rows.
+#[derive(Debug, Default)]
+pub struct RowMerger {
+    /// The row's current state. State progresses while processing a single chunk.
+    state: ReadState,
+    /// The last row key we've encountered. This should be consistent across the chunks.
+    last_seen_row_key: Option<RowKey>,
+    /// The last cell family. This may change, indicating a new cell group.
+    last_seen_cell_family: Option<FamilyId>,
+    /// The row that is currently being compiled.
+    row_in_progress: RefCell<PartialRow>,
+}
+
+impl RowMerger {
+    /// discard data so far and return to a neutral state.
+    async fn reset_row(
```

Comment:
I think the only method in here that needs async is `process_chunks`

```suggestion
    fn reset_row(
```